connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTaskThread.java

```diff
@@ -50,16 +50,20 @@ public WorkerSinkTaskThread(WorkerSinkTask task, String name, Time time,
 
     @Override
     public void execute() {
-        // Try to join and start. If we're interrupted before this completes, bail.
-        if (!task.joinConsumerGroupAndStart())
-            return;
+        try {
+            // Try to join and start. If we're interrupted before this completes, bail.
+            if (!task.joinConsumerGroupAndStart())
+                return;
 
-        while (getRunning()) {
-            iteration();
-        }
+            while (getRunning()) {
+                iteration();
+            }
 
-        // Make sure any uncommitted data has committed
-        task.commitOffsets(true, -1);
+            // Make sure any uncommitted data has committed
+            task.commitOffsets(true, -1);
+        } catch (Throwable t) {
+            task.abort(t);
```

Comment:
Yeah, I agree we eventually want to do some refactoring to make the job of each piece, and the state it owns, clearer.

Re: the particular issue here, I guess I'm now confused about what you're proposing. Running `abort()` like it is in the patch right now is definitely not safe since then `task.stop()` might be invoked twice, right? I think what you're getting at is that even before this patch, the thread could exit and we wouldn't really do anything in the `Worker` -- `ShutdownableThread.run()` would log an error with the exception but then we wouldn't really do anything to clean up the task until something else triggered the `Worker` to clean it up. This will happen eventually (due to rebalance or shutdown), but isn't really ideal.

Does adding the `abort` here in this patch actually accomplish anything? It would shut down the task sooner, but all the state associated with it in the `Worker` wouldn't be cleaned up (or minimized, since we probably still want to track the fact that we own the task).

If we *did* want to clean up earlier in this thread, would it make more sense to add the relevant callback that we're going to need eventually anyway? If we want this component to clean up the task when it fails, then we need to track the fact that we did that somewhere so that any subsequent cleanup (`WorkerSinkTask.stop()` and `WorkerSinkTask.awaitStop()`) should skip that step since it's already been performed. Or, if the callback were provided, the full cleanup of the task, using the normal mechanism, could be implemented (presumably in code in Worker or something, although I'm not sure where we want that callback routed just yet since it probably depends on where we track the status info).